mittab/apps/tab/pairing_views.py

```diff
@@ -105,6 +107,100 @@ def assign_judges_to_pairing(request):
                                             "Got error during judge assignment")
     return redirect("/pairings/status/")
 
+@permission_required("tab.tab_settings.can_change", login_url="/403/")
+def assign_rooms_to_pairing(request):
+    current_round_number = TabSettings.objects.get(key="cur_round").value - 1
+    warnings = []
+    if request.method == "POST":
+        try:
+            # A bit ambivalent about including a backup. Its probably good
+            # But the backup list might start getting long
+            backup.backup_round("round_%s_before_room_assignment" %
+                                current_round_number)
+            warnings = assign_rooms.add_rooms()
+            request.session['room_warnings'] = warnings
+        except Exception:
+            emit_current_exception()
+            return redirect_and_flash_error(request,
+                                            "Got error during room assignment")
+    return redirect("/pairings/status/")
+
+def alternative_rooms(request, round_id, room_id=""):
+    round_obj = Round.objects.get(id=int(round_id))
+    round_number = round_obj.round_number
+
+    current_room_obj = None
+    if room_id.isdigit():
+        try:
+            current_room_obj = Room.objects.get(id=int(room_id))
+        except Room.DoesNotExist:
+            pass
+
+    # Fetch all rooms checked in for the given round, ordered by rank
+    # A little out of depth on the Django ORM here, but this seems like the
+    # way to get this data with the least amount of queries
+    rooms = set(Room.objects.filter(
+        roomcheckin__round_number=round_number
+    ).annotate(
+        has_round=Exists(Round.objects.filter(room_id=OuterRef('id')))
+    ).order_by("-rank"))
```

Comment:
The `set()` is for the subtraction operator, right?

I'd prefer something more like:

```python3
viable_unpaired_rooms = []
viable_paired_rooms = []
other_rooms = [] # populate this if needed, otherwise delete

for room in rooms:
    if room.has_round:
        viable_paired_rooms.append(room)
    else:
        viable_unpaired_rooms.append(room)
```